### PR TITLE
Add code for sentiment analysis

### PR DESCRIPTION
**Pull request by MSanKeys963**
Removed `class NLPHandler()` and added sentiment analysis functionality in `class MLHandler()`.

To setup a Gramex service for performing sentiment analysis, use the following configuration:

```yaml
url:
  sentiment-analysis:
    pattern: /$YAMLURL/
    handler: MLHandler
    kwargs:
      backend: transformers
      task: sentiment-analysis
      xsrf_cookies: false
```

## Getting predictions

`GET` sentiments of short pieces of text as follows:

```bash
curl -X GET --data-urlencode "text=This movie is so bad, it's good." http://localhost:9988/
```

The output will be:
```json
[
  {
    "label": "POSITIVE",
    "score": 0.9997316002845764
  }
]
```

Files containing text to be classified can also be `POST`ed to the endpoint, with `_action=predict`. Any file supported by `gramex.cache.open` will work. (Download a sample [here](https://drive.google.com/file/d/146MXfbBNpC8xI0Txpra3xs25WHpHk73k/view?usp=sharing).)

```bash
curl -X POST -F "file=@sentiment.csv" http://localhost:9988/?_action=predict
```

The output will be:

```json
[
  {
    "label": "POSITIVE",
    "score": 0.9997316002845764
  },
  {
    "label": "NEGATIVE",
    "score": 0.9974692463874817
  },
  // etc.
]
```

## Measuring model performance

Files containing the `text` and `label` fields can be `POSTED` to the endpoint
with `_action=score` to get the [ROC AUC score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html) of the model against the dataset. (Download a sample dataset [here](https://drive.google.com/file/d/146MXfbBNpC8xI0Txpra3xs25WHpHk73k/view?usp=sharing)).

```bash
curl -X POST -F "file=@sentiment.csv" http://localhost:9988/?_action=score
```

The output will be something like:

```json
{
  "roc_auc": 0.9929
}
```

## Training the model

The model can be trained on a dataset by setting `_action=train`, and `POST`ing the file.

```bash
curl -X POST -F "file=@sentiment_score.json" http://localhost:9988/?_action=train
```

The output will show the score of the trained model on the dataset:

```json
{
  "roc_auc": 0.8
}
```

Multiple training options for the transformer are supported, including the number of epochs, batch size and weight decay. These can all be specified in the `POST` request as follows:

```bash
# Train for three epochs instead of the default 1
curl -X POST -F "file=@sentiment.csv" http://localhost:9988/?_action=train&num_train_epochs=3
```

The output is the score of the trained model on the dataset after 3 epochs:

```json
{
  "roc_auc": 0.98
}
```

```bash
# Change the batch size to 32 instead of the default 16
curl -X POST -F "file=@sentiment.csv" \
	http://localhost:9988/?_action=train&per_device_train_batch_size=32&num_train_epochs=3
```

The output is the score of the trained model on the dataset after 3 epochs and a batch size of 32:

```json
{
  "roc_auc": 0.99
}
```

### CODE REVIEW

**Comment by sanand0** [comment]
Cool! @jaidevd could you please review? Do let me know when to merge

**Comment by jaidevd** [comment]
@MSanKeys963 The target branch has to be `gramener/gramex`'s master branch, not the `jd-transformers` branch.

File: gramex/handlers/mlhandler.py [hunk 1]
```diff
@@ -450,6 +450,10 @@ def get(self, *path_args, **path_kwargs):
             self.write(json.dumps(params, indent=2))
         elif '_cache' in self.args:
             self.write(self.load_data().to_json(orient='records'))
+        elif 'text' in self.args:
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:453]: This condition doesn't check `backend`. Which means, that even if I'm running an sklearn model which has one column named `text`, and I request predictions from this model - this piece of code will run.

File: gramex/handlers/mlhandler.py [hunk 2]
```diff
@@ -262,13 +264,11 @@ def setup(cls, data=None, model={}, backend='sklearn', config_dir='', **kwargs):
         # elif backend == 'transformers':
         #     NLPHandler.fit(**kwargs)
         if backend != 'sklearn':
+            cls.model_backend = backend
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:267]: Do we need this extra variable? `backend` can imply be assigned to `cls.backend`.

**Review by jaidevd — CHANGES_REQUESTED** [review]

**Comment by jaidevd** [comment]
@MSanKeys963 other than these two changes, LGTM

**Comment by jaidevd** [comment]
@MSanKeys963 this still showing merge conflicts. Please take a look.

File: gramex/handlers/formhandler.py [hunk 3]
```diff
@@ -53,6 +53,7 @@ def setup(cls, **kwargs):
         cls.headers = conf_kwargs.pop('headers', {})
         # Top level formats: key is special. Don't treat it as data
         cls.formats = conf_kwargs.pop('formats', {})
+        cls.task = conf_kwargs.pop('task', {})
```
Comment (jaidevd) [inline gramex/handlers/formhandler.py:56]: @MSanKeys963 We should not have to make any changes to FormHandler.

File: gramex/install.py [hunk 4]
```diff
@@ -197,32 +197,42 @@ def _ensure_remove(function, path, exc_info):
...
+            # auto-resuming rmtree: https://bugs.python.org/issue8523
+            _try_remove(target, retries, delay, func, **kwargs)
+            return True
+        else:
+            app_log.warning('Not removing directory %s (outside $GRAMEXDATA)', target)
+            return False
     else:
-        app_log.warning('Not removing directory %s (outside $GRAMEXDATA)', target)
-        return False
+        _try_remove(target, retries, delay, func, **kwargs)
```
Comment (jaidevd) [inline gramex/install.py:236]: This file should not have changed. Please replace this file with the version available on master.

**Review by jaidevd — CHANGES_REQUESTED** [review]

File: gramex/handlers/formhandler.py (hunk 3 above)
Comment (MSanKeys963) [reply gramex/handlers/formhandler.py:56]: Can we keep this line?
It takes care of `task: sentiment-analysis` from `gramex.yaml`.

Comment (jaidevd) [reply gramex/handlers/formhandler.py:56]: Why do we need `cls.task` in FormHandler? It doesn't mean anything to the FormHandler class. This attribute is just left here, and the class doesn't do anything with it. Also, if it's just a matter of popping the task key from the `conf_kwargs` dict, that can easily be done in `MLHandler` itself, before calling `super(MLHandler, ...)`

File: gramex/handlers/mlhandler.py [hunk 5]
```diff
@@ -369,9 +452,29 @@ def post(self, *path_args, **path_kwargs):
...
+            elif action == 'predict':
+                args = self.model, data['text'].tolist()
+            res = yield gramex.service.threadpool.submit(*args, **kwargs)
+            self.write(json.dumps(res, indent=2, cls=CustomJSONEncoder))
+        else:
+            res = yield gramex.service.threadpool.submit(getattr(self, f"_{action}"))
+            self.write(json.dumps(res, indent=2, cls=CustomJSONEncoder))      
         super(MLHandler, self).post(*path_args, **path_kwargs)
+            
+    def get_cached_arg(self, argname):
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:474]: It turns out that we don't need this function either. It's not being used anywhere. Please remove this as well.

File: gramex/handlers/formhandler.py (hunk 3 above)
Comment (MSanKeys963) [reply gramex/handlers/formhandler.py]: Got it. Thank you!

**Comment by MSanKeys963** [comment]
@jaidevd I've fixed all the issues mentioned above. Please let me know if there's anything else.

**Comment by jaidevd** [comment]
Thanks, @MSanKeys963 

@sanand0 This is ready for merge.

File: gramex/dl_utils.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from torch.utils.data import Dataset
```
Comment (sanand0) [inline gramex/dl_utils.py:1]: Let's rename this file to `dl.py`

File: gramex/handlers/mlhandler.py [hunk 7]
```diff
@@ -16,11 +16,20 @@
...
 from sklearn.pipeline import Pipeline
-from sklearn.preprocessing import OneHotEncoder, StandardScaler
+from sklearn.metrics import roc_auc_score
+from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
 from slugify import slugify
 from tornado.gen import coroutine
 from tornado.web import HTTPError
 from sklearn.metrics import get_scorer
+try:
+    from transformers import pipeline
```
Comment (sanand0) [inline gramex/handlers/mlhandler.py:26]: Let's have a consistent policy for imports. Treat torch and huggingface the same way.

Comment (MSanKeys963) [reply gramex/handlers/mlhandler.py]: @sanand0  Fixed it.

**Comment by MSanKeys963** [comment]
@sanand0 I've fixed all the issues. Please check.

**Comment by sanand0** [comment]
@MSanKeys963

- [ ] Can you get this to work, please? [sentiment.zip](https://github.com/gramener/gramex/files/6905106/sentiment.zip)
- [ ] Gramex should still run if PyTorch & Huggingface are not installed

For example, this is how we optionally import ElasticSearch:

```python
def gramexlog(conf):
    try:
        from elasticsearch import Elasticsearch, helpers
    except ImportError:
        app_log.error('gramexlog: elasticsearch missing. pip install elasticsearch')
        return
```
